integration_tests/src/main/python/parquet_pyarrow_test.py

```diff
@@ -0,0 +1,255 @@
+    # convert py_dict to Pyspark Row
+    if py_dict is None:
+        return None
+    else:
+        return Row(**py_dict)
+
+
+def _convert_pyarrow_list_object_to_cpu_object(pyarrow_list_scalar):
+    """Convert pyarrow list to python list"""
+    # pyarrow.ListScalar obj to python list
+    py_list = pyarrow_list_scalar.as_py()
+    if py_list is None:
+        return None
+    else:
+        return py_list
+
+
+def _convert_pyarrow_map_object_to_cpu_object(pyarrow_map_scalar):
+    """Convert pyarrow map to python dict"""
+    # pyarrow.MapScalar obj to python (key, value) pair list
+    py_key_value_pair_list = pyarrow_map_scalar.as_py()
+    if py_key_value_pair_list is None:
+        return None
+    else:
+        return dict(py_key_value_pair_list)
+
+
+def _convert_pyarrow_column_to_cpu_object(pyarrow_column, pa_type):
+    """ convert pyarrow column to a column contains CPU type objects"""
+    if isinstance(pa_type, pa.StructType):
+        # struct column
+        return [_convert_pyarrow_struct_object_to_cpu_object(v) for v in pyarrow_column]
+    elif isinstance(pa_type, pa.ListType):
+        # list column
+        return [_convert_pyarrow_list_object_to_cpu_object(v) for v in pyarrow_column]
+    elif isinstance(pa_type, pa.MapType):
+        # map column
+        return [_convert_pyarrow_map_object_to_cpu_object(v) for v in pyarrow_column]
+    else:
+        # other type column
+        return [v.as_py() for v in pyarrow_column]
+
+
+def _gen_row_from_columnar_cpu_data(column_names, columnar_data, row_idx):
+    """Gen a Pyspark Row from columnar data for a row index"""
+    key_value_pair_list = [(column_names[col], columnar_data[col][row_idx]) for col in range(len(column_names))]
+    return Row(**dict(key_value_pair_list))
+
+
+def _convert_pyarrow_table_to_cpu_object(pa_table):
+    """
+    Convert pyarrow table to cpu type data to reuse the `_assert_equal(cpu, gpu)` in asserts.py
+     e.g.: convert pyarrow map_scalar to Pyspark Row
+    """
+    columnar_data = []
+    for pa_column, pa_type in zip(pa_table.columns, pa_table.schema.types):
+        converted_column = _convert_pyarrow_column_to_cpu_object(pa_column, pa_type)
+        columnar_data.append(converted_column)
+    cpu_data = [
+        _gen_row_from_columnar_cpu_data(pa_table.schema.names, columnar_data, row_idx)
+        for row_idx in range(pa_table.num_rows)
+    ]
+    return cpu_data
+
+
+def assert_equal(cpu_adapted_from_pyarrow, gpu):
+    """Verify that the result from the pyarrow(already adapted to CPU data format) and the GPU are equal"""
+    try:
+        _assert_equal(cpu_adapted_from_pyarrow, gpu, float_check=get_float_check(), path=[],
+                      cpu_data_adapted_from="pyarrow")
+    except:
+        sys.stdout.writelines(difflib.unified_diff(
+            a=[f"{x}\n" for x in cpu_adapted_from_pyarrow],
+            b=[f"{x}\n" for x in gpu],
+            fromfile='pyarrow OUTPUT',
+            tofile='GPU OUTPUT'))
```

Comment:
Related to the "is this necessary" comment above, not a fan of replicating code here.

The asserts module already has the ability to check reads and writes.  The only reason these cannot be used as-is for pyarrow testing is that they don't allow different read/write functions to be used for CPU vs. GPU.  If we refactored the code to have those read/write assert functions take separate CPU vs  GPU functions (and make the existing APIs leverage those by passing the same function for both CPU and GPU) then it would be straightforward to add pyarrow wrappers around the assert code in the pyarrow module without replicating code or needing to import low-level assert functions (i.e.: anything starting with an underscore, which indicates this is not intended to be used by other modules).

Reply:
Done.